tree/ntuple/v7/inc/ROOT/RNTupleMetrics.hxx

```diff
@@ -0,0 +1,253 @@
+/// \file ROOT/RNTupleMetrics.hxx
+/// \ingroup NTuple ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2019-08-27
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2019, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RNTupleMetrics
+#define ROOT7_RNTupleMetrics
+
+#include <TError.h>
+
+#include <atomic>
+#include <chrono>
+#include <cstdint>
+#include <ctime> // for CPU time measurement with clock()
+#include <memory>
+#include <ostream>
+#include <string>
+#include <vector>
+#include <utility>
+
+namespace ROOT {
+namespace Experimental {
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTuplePerfCounter
+\ingroup NTuple
+\brief A performance counter with a name and a unit, which can be activated on demand
+
+Derived classes decide on the counter type and implement printing of the value.
+*/
+// clang-format on
+class RNTuplePerfCounter {
+private:
+   std::string fName;
+   std::string fUnit;
+   std::string fDescription;
+   bool fIsEnabled = false;
+
+public:
+   RNTuplePerfCounter(const std::string &name, const std::string &unit, const std::string &desc)
+      : fName(name), fUnit(unit), fDescription(desc) {}
+   void Enable() { fIsEnabled = true; }
+   bool IsEnabled() const { return fIsEnabled; }
+   std::string GetName() const { return fName; }
+   std::string GetDescription() const { return fDescription; }
+   std::string GetUnit() const { return fUnit; }
+
+   virtual std::string ToString() const = 0;
+};
+
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RNTuplePlainCounter
+\ingroup NTuple
+\brief A non thread-safe integral performance counter
+*/
+// clang-format on
+class RNTuplePlainCounter : public RNTuplePerfCounter {
+private:
+   std::int64_t fCounter = 0;
+
+public:
+   RNTuplePlainCounter(const std::string &name, const std::string &unit, const std::string &desc)
+      : RNTuplePerfCounter(name, unit, desc)
+   {
+   }
+
+   void Inc() { ++fCounter; }
+   void Dec() { --fCounter; }
+   void Add(int64_t delta) { fCounter += delta; }
+   int64_t GetValue() const { return fCounter; }
+   void SetValue(int64_t val) { fCounter = val; }
+   std::string ToString() const override { return std::to_string(fCounter); }
```

Comment:
Shouldn't it also include the unit (and maybe the name)?

Reply:
Hm.. the printing method eventually used is the one in `Metrics`, which adds unit and name. We could move that string concatenation in the Counter class itself, but then we'd need to repeat the pattern for all the counter variations. Not sure what's better.

Reply:
ok. the way you have is probably fine.  An alternative (amonst others) is to have, per se, 2 ToString routine.  One (ValueToString) that is overload in the derived class and does just the value and the one in the base class that assemble the 3 things and is not virtual.